[SecuritySolution] Add Privileged user activity dashboards to Privmon page

machadoum opened the pull request:
## Summary

** Changes to the Privimon page are behind the experimental flag `privilegeMonitoringEnabled`

Add the privileged user activity panel to the privmon dashboard page.
The user activity panel has three visualisations that switch places when a button group is clicked.
When that happens, we rerender the embedded visualisation with different columns, data and lens attributes.


### Not implemented
- [ ] Method column. We were unable to find data to fill this column correctly.
- [x] Add instructions to generate data. We don't have a data generator yet.
- [ ] When filtering by a value, we should add an OR of all fields (because data comes from different fields)
- [ ] Index not found and field not found errors


### How to test it
* Enable the experimental flag `privilegeMonitoringEnabled`
* Start Kibana and ES
* Use the document generator to generate data
  * On the [branch](https://github.com/elastic/security-documents-generator/tree/privileged_user_monitoring_visualizations) run `yarn start privileged_user_monitoring`
* On Kibana Dev tools run `POST kbn:/api/entity_analytics/monitoring/engine/init`
* Open the Entity analytics page and navigate to dashboards
* The Privileged user activity dashboards should be empty because there is no privileged user 
* Add privileged users to the index using the dev console
```
POST kbn:/api/entity_analytics/monitoring/users
{
  "user": {"name": "john.smith"}
}
POST kbn:/api/entity_analytics/monitoring/users
{
  "user": {"name": "stacy_armstrong"}
}
POST kbn:/api/entity_analytics/monitoring/users
{
  "user": {"name": "john_smith"}
}
POST kbn:/api/entity_analytics/monitoring/users
{
  "user": {"name": "randy.carlisle"}
}
POST kbn:/api/entity_analytics/monitoring/users
{
  "user": {"name": "root"}
}
``` 
* Open the Entity analytics page and navigate to dashboards
* Now, the Privileged user activity dashboards should display data



### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

machadoum committed:
wip

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

machadoum committed:
Add Privileged user activity dashboards to Privimon page

machadoum committed:
Merge branch 'main' into siem-ea-privmon-dashboards-2

machadoum committed:
Add discover link

machadoum committed:
Self code review

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

machadoum committed:
Merge branch 'main' into siem-ea-privmon-dashboards-2

machadoum committed:
Fix build

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/303015)
* Commit: 11466dd53508fa0fcb2c8833f580796239506415

### Failed CI Steps
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-9498-4575-aed4-3fe272806e98)
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716ce-3d32-4496-8991-4da05e301afa)
* [FTR Configs #14](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94a7-46ff-bbf0-6ed948b0a5e3)
* [FTR Configs #15](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94a8-4470-80b3-967a3ef7921c)
* [FTR Configs #41](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94b5-47b2-8899-093d1cbc7d56)
* [FTR Configs #55](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94bc-4346-a35e-be6c88f6be9b)
* [FTR Configs #72](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94c5-4d91-a7a2-834bf14b07a0)
* [FTR Configs #73](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94c6-4081-a487-6b293abc179a)
* [FTR Configs #125](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94e5-4663-b528-853c61e5ccb4)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-9498-4575-aed4-3fe272806e98) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716a7-9498-4575-aed4-3fe272806e98/artifacts/019716ce-1750-4864-86b5-75b24d57fc82) Jest Tests #<span></span>1 / columns buildAuthenticationsColumns renders destination column
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716ce-3d32-4496-8991-4da05e301afa) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716ce-3d32-4496-8991-4da05e301afa/artifacts/019716d8-1957-4e7b-8179-5b5889544859) Jest Tests #<span></span>1 / columns buildAuthenticationsColumns renders destination column
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-9498-4575-aed4-3fe272806e98) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716a7-9498-4575-aed4-3fe272806e98/artifacts/019716ce-1750-447d-ac28-877590789e68) Jest Tests #<span></span>1 / columns buildAuthenticationsColumns renders ip column
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716ce-3d32-4496-8991-4da05e301afa) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716ce-3d32-4496-8991-4da05e301afa/artifacts/019716d8-1957-4d2b-9d2d-e8e77470406c) Jest Tests #<span></span>1 / columns buildAuthenticationsColumns renders ip column
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-9498-4575-aed4-3fe272806e98) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716a7-9498-4575-aed4-3fe272806e98/artifacts/019716ce-1750-4d8e-895b-1467a25070c4) Jest Tests #<span></span>1 / columns buildAuthenticationsColumns renders result column with badge
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716ce-3d32-4496-8991-4da05e301afa) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716ce-3d32-4496-8991-4da05e301afa/artifacts/019716d8-1956-4782-86a2-e0e1155dd072) Jest Tests #<span></span>1 / columns buildAuthenticationsColumns renders result column with badge
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-9498-4575-aed4-3fe272806e98) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716a7-9498-4575-aed4-3fe272806e98/artifacts/019716ce-1750-458f-b2e6-2f4108fc770a) Jest Tests #<span></span>1 / esql_data_generation generateListESQLQuery should generate the correct ESQL query for list pagination
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716ce-3d32-4496-8991-4da05e301afa) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716ce-3d32-4496-8991-4da05e301afa/artifacts/019716d8-1957-46d1-a1ee-86ca618853d1) Jest Tests #<span></span>1 / esql_data_generation generateListESQLQuery should generate the correct ESQL query for list pagination
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/303015#019716a7-94a8-4470-80b3-967a3ef7921c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/303015/jobs/019716a7-94a8-4470-80b3-967a3ef7921c/artifacts/019716d1-bf90-4c24-a6e0-9367aacaa62c) FTR Configs #<span></span>15 / Observability AI Assistant Functional tests settings/change_knowledge_base_model.spec.ts Change knowledge base model via settings should show a button to re-deploy the model if the model has been stopped

### History
* :broken_heart: [Build #302978](https://buildkite.com/elastic/kibana-pull-request/builds/302978) failed d0d7c3c0fb92d5de9fdd3fdbfbf3ad7c7b1101bb

cc @machadoum

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303015","number":303015,"commit":"11466dd53508fa0fcb2c8833f580796239506415"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/302978","number":302978,"commit":"d0d7c3c0fb92d5de9fdd3fdbfbf3ad7c7b1101bb"}],"number":303015}
buildkite-pr-comment-->

machadoum committed:
Fix unit test

machadoum committed:
Update esql query

machadoum committed:
Improve the type column for the auth dashboard

jaredburgettelastic commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/esql_source_query.ts:
```diff
@@ -0,0 +1,49 @@
+const PRIV_MON_JOIN = ``;
+
+export const getGrantedRightsEsqlSource = () => {
+  return `FROM logs-*, test* METADATA _id, _index
```
Is the `test*` index pattern intentional here?

machadoum replied:
No. 😓

machadoum committed:
Fix ESQL queries

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/303600)
* Commit: a0497fffa6377e3a38cff33ee488fc2f8ff5b101

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/aeb129b30ca3f70a319bfc5b3230a21b7d36ee91) | [after](https://github.com/elastic/kibana/commit/a0497fffa6377e3a38cff33ee488fc2f8ff5b101) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 7436 | 7447 | +11 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/aeb129b30ca3f70a319bfc5b3230a21b7d36ee91) | [after](https://github.com/elastic/kibana/commit/a0497fffa6377e3a38cff33ee488fc2f8ff5b101) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.2MB | 9.2MB | +13.4KB |



### History
* :broken_heart: [Build #303133](https://buildkite.com/elastic/kibana-pull-request/builds/303133) failed 24999ff86606a281928dd283e54774e9d37e3936
* :broken_heart: [Build #303015](https://buildkite.com/elastic/kibana-pull-request/builds/303015) failed 11466dd53508fa0fcb2c8833f580796239506415
* :broken_heart: [Build #302978](https://buildkite.com/elastic/kibana-pull-request/builds/302978) failed d0d7c3c0fb92d5de9fdd3fdbfbf3ad7c7b1101bb

cc @machadoum

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303600","number":303600,"commit":"a0497fffa6377e3a38cff33ee488fc2f8ff5b101"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303133","number":303133,"commit":"24999ff86606a281928dd283e54774e9d37e3936"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303015","number":303015,"commit":"11466dd53508fa0fcb2c8833f580796239506415"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/302978","number":302978,"commit":"d0d7c3c0fb92d5de9fdd3fdbfbf3ad7c7b1101bb"}],"number":303600}
buildkite-pr-comment-->

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-privmon-dashboards-2

machadoum committed:
Makde several priv mon user fix and improvements

Add extra null checks for the columns
Add namespace support for privileged user queries
Rename privileged user index for all dashboard queries
Change privileged user index mode to lookup

hop-dev commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/columns.tsx:
```diff
@@ -0,0 +1,304 @@
+    user != null
+      ? getRowItemsWithActions({
+          values: isArray(user) ? user : [user],
+          // TODO can we have an array of field names and do an OR?
```
I would prefer to have a tech debt issue for this so it doesn't get lost

machadoum replied:
We have one. https://github.com/elastic/security-team/issues/12712
I will update the comment to link to the issue

hop-dev commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/columns.tsx:
```diff
@@ -0,0 +1,304 @@
+      defaultMessage="Timestamp"
+    />
+  ),
+  width: '20%',
```
An idea, maybe a bad one,  could we declare a const object with all of the column width percentage values together so that its easy to see the differences/tweak them? 

const COLUMN_WIDTHS = { TIMESTAMP: '20%', ...

hop-dev commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/columns.tsx:
```diff
@@ -0,0 +1,304 @@
+    },
+    truncateText: true,
+  },
+  // {
```
Whats the story here? Tech debt or something that needs fixing in this PR?

machadoum commented on x-pack/solutions/security/plugins/security_solution/public/entity_analytics/components/privileged_user_monitoring/components/privileged_user_activity/columns.tsx:
```diff
@@ -0,0 +1,304 @@
+        return getEmptyTagValue();
+      }
+
+      if (type.startsWith('/api/v1/authn')) {
```
Good one! We probably need to improve this logic. The investigation is also part of this ticket https://github.com/elastic/security-team/issues/12713

machadoum committed:
Code review improvements

hop-dev reviewed (approved):
Code review only 👍

CAWilson94 commented:
Desk tested, looking good! Thank you! 🚀 

###  ✅ 'Privileged user activity' Section, Confirmed Receiving Data 
All looking good! 🥳 

![image](https://github.com/user-attachments/assets/54ed1036-7bc8-43d3-b49c-107ca6f7c234)

![image](https://github.com/user-attachments/assets/349f5a2b-5681-45ad-b762-3e236211333f)

![image](https://github.com/user-attachments/assets/2d6868db-aef1-413b-b1ca-42fbe5e2510e)

machadoum committed:
Fix unit tests

kapral18 commented on x-pack/solutions/security/plugins/security_solution/public/common/components/visualization_actions/use_lens_attributes.tsx:
```diff
@@ -123,7 +119,28 @@ export const useLensAttributes = ({
-    const query = esqlQuery ? { esql: esqlQuery } : globalQuery;
+    const query = esql ? { esql } : globalQuery;
+
+    const queryFilters =
+      applyGlobalQueriesAndFilters && esql
+        ? [
+            {
+              meta: {
+                alias: null,
+                disabled: false,
+                key: 'globalFilterKey',
+                negate: false,
+                params: {},
+                type: 'string',
+              },
+              query: {
+                bool: {
+                  filter: globalFilterQuery,
+                },
+              },
+            },
+          ]
+        : [...(applyGlobalQueriesAndFilters ? filters : [])];
```
NIT: Nested ternaries are harder to read, wdyt about:

```suggestion
const queryFilters = (() => {
  if (!applyGlobalQueriesAndFilters) return [];
  
  if (esql) {
    return [{
      meta: {
        alias: null,
        disabled: false,
        key: 'globalFilterKey',
        negate: false,
        params: {},
        type: 'string',
      },
      query: {
        bool: {
          filter: globalFilterQuery,
        },
      },
    }];
  }
  
  return filters
})();
```

machadoum replied:
Thank you for the comment!

kapral18 reviewed (approved):
LGTM for explore, 1 comment for improvement
thanks

machadoum committed:
Improve SQLGlobalFilters code based on code review

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-privmon-dashboards-2

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/306191)
* Commit: 8b3e129795faca04d9f1d7bf32d5f0b2593de36f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/f8d4ae31c41a97c4fc799b7a54d0548ff9d635e3) | [after](https://github.com/elastic/kibana/commit/8b3e129795faca04d9f1d7bf32d5f0b2593de36f) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 7574 | 7587 | +13 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f8d4ae31c41a97c4fc799b7a54d0548ff9d635e3) | [after](https://github.com/elastic/kibana/commit/8b3e129795faca04d9f1d7bf32d5f0b2593de36f) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.3MB | 9.3MB | +13.7KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/f8d4ae31c41a97c4fc799b7a54d0548ff9d635e3) | [after](https://github.com/elastic/kibana/commit/8b3e129795faca04d9f1d7bf32d5f0b2593de36f) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 93.2KB | 93.2KB | -1.0B |



### History
* :yellow_heart: [Build #305280](https://buildkite.com/elastic/kibana-pull-request/builds/305280) was flaky 473255ac83d5a69ffec2dceda6e2841d74ab0cff
* :broken_heart: [Build #305204](https://buildkite.com/elastic/kibana-pull-request/builds/305204) failed 240fe8d1e5b73961cdf8b0748f9e2ed35980e309
* :broken_heart: [Build #305154](https://buildkite.com/elastic/kibana-pull-request/builds/305154) failed 9a2e43645b3523b9611d64a38119783001f9969a
* :green_heart: [Build #303600](https://buildkite.com/elastic/kibana-pull-request/builds/303600) succeeded a0497fffa6377e3a38cff33ee488fc2f8ff5b101
* :broken_heart: [Build #303133](https://buildkite.com/elastic/kibana-pull-request/builds/303133) failed 24999ff86606a281928dd283e54774e9d37e3936
* :broken_heart: [Build #303015](https://buildkite.com/elastic/kibana-pull-request/builds/303015) failed 11466dd53508fa0fcb2c8833f580796239506415

cc @machadoum

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/306191","number":306191,"commit":"8b3e129795faca04d9f1d7bf32d5f0b2593de36f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305280","number":305280,"commit":"473255ac83d5a69ffec2dceda6e2841d74ab0cff"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305204","number":305204,"commit":"240fe8d1e5b73961cdf8b0748f9e2ed35980e309"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/305154","number":305154,"commit":"9a2e43645b3523b9611d64a38119783001f9969a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303600","number":303600,"commit":"a0497fffa6377e3a38cff33ee488fc2f8ff5b101"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/303133","number":303133,"commit":"24999ff86606a281928dd283e54774e9d37e3936"}],"number":306191}
buildkite-pr-comment-->
